### ISSUE TITLE
DOM / Web APIs

### ISSUE THREAD

**Issue by kieranpotts** [craftingpress/crafting-websites#10]
Event attributes
================

Event attributes (onclick, onhover, etc.) must not be used – ever. Callback functions must be applied to DOM elements dynamically.

<data> elements and data-* attributes
=====================================

Traditionally we have used class and id attributes as targets for DOM queries.

<a href="/login.html" class="login">Login</a>

...

var login_link = document.querySelector( '.login' );

var popup = new Popup( login_link.href, { width: 500, modal: true } );

login_link.addEventListener( 'click', popup.show, false );

This works well. But it does not provide any standardised means of providing input data to our JavaScript programs. In the example above, the width and format of the "Login" popup window have to be configured via inline scripting. And classes are taking on dual responsibilities....

The new <data> element and data-* attributes provide us with standard, unobtrusive markup for passing data from HTML to JavaScript. We should use them as the exclusive means of providing input to our JavaScript applications.

<a href="/login.html" data-popup="true" data-popup-width="750">Login</a>

...

var login_link = document.querySelector( '[data-popup="true"]' );

var width = Number( login_link.getAttribute( 'data-popup-width' ) || 450 );

var popup = new Popup( login_link.href, { width: width, modal: true } );

login_link.addEventListener( 'click', popup.show, false );

This approach to developing JavaScript plugins for web interfaces means that we can do away with inline scripting altogether.

Remember, everything that is initially rendered in the page must work without error, even if JavaScript enhancements fail to load. In the example above, the "Login" link will redirect users to the standard login page. For some users, however, scripting improves the experience by renderingh the login form in a popup window. And even for users with JavaScript running in their browser, this unobtrusive way of applying dynamic behaviours to the link does not break the right-click "Open in new tab" functionality.

As well as data-* attributes, we can also use <data> elements to provide equivalent machine-readable for bits of information that are printed in a human-friendly format. The <time> element is also useful for this purpose, though it is less flexible than the <data> element because the visible value is expected to be in a machine-readable format.

<ul>

<li>Where: <data id="location" value="546">Adventure Cade</data></li>

<li>When: <time id="start">2014-12-01 07:00</time></li>

<li>Duration: <data id="duration" value="1:45">1hr 45mins</data></li>

</ul>

Incidentally the <data> element can also be used in conjunction with microformats, microdata and RDFa syntax to integrate metadata in our web pages. In the following example some HTML parsers will recognise the "dtstart" and "dtend" classes from the hCalendar syntax, part of the microformats specification. A web browser or third party web application could make use of this, for example by allowing the user to add the event to a calendar – at the click of a button. But as a general rule: data elements and attributes should provide data to the internal application; metadata markip is for providing data to external applications.

<data class="dtstart" value="2014-12-01">1st</data>

– <data class="dtend" value="2014-12-10">10th</data> December

Officially there is practically zero support for the <data> element among the current crop of browsers. No matter. We can still use it. JavaScript has no problem identifying this element and reading its attributes.

I recommend using data elements and attributes and the exclusive link between HTML and JavaScript, for the sake of consistency. They should even be the target of our DOM queries.

var login_link = document.querySelector( '*[data-popup="true"]' ); 

Think of data elements and attributes for JavaScript as the equivalent of classes for CSS. Use data elements and attributes to separate logic and functionalityt from HTML content. Use class attributes to separate content from its presentation.

xxxxx

x

It has long been considered bad practice (and computationally expensive) to manually create large blocks of HTML markup in-memory through string concatenation. Developers using this technique often find themselves iterating through their data, wrapping it in nested divs and using outdated techniques such as document.write to inject the ‘template’ into the DOM. This approach often means keeping scripted markup inline with standard markup, which can quickly become difficult to read and maintain, especially when building large applications.

Querying the DOM

xxx

HTMLFormElement.elements

https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/elements

// Firefox includes fieldsets inside elements nodelist so we filter it out.

var _inputs = this._form.elements;

var _length = _inputs.length;

var _input;

for ( var _i = 0; _i < _length; _i++ ) {

 _input = _inputs[ _i ];

 if ( /^(submit|image|button|reset)$/i.test( _input.type ) ) {

Cloning and creating new DOM

https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment

https://developer.mozilla.org/en-US/docs/Web/API/document/createDocumentFragment

http://stackoverflow.com/questions/3397161/

NOT well supported yet: the DOMParser interface -

https://developer.mozilla.org/en-US/docs/Web/API/DOMParser

https://developer.mozilla.org/en-US/Add-ons/Code_snippets/HTML_to_DOM

http://caniuse.com/#search=domparser

Properties and attributes

http://stackoverflow.com/questions/6003819/properties-and-attributes-in-html

http://stackoverflow.com/questions/8018814/setting-a-property-via-property-or-setattribute

http://stackoverflow.com/questions/3919291/when-to-use-setattribute-vs-attribute-in-javascript

http://stackoverflow.com/questions/11886820/what-is-the-difference-between-setattribute-and-dot-notation-in-javascript

Getting the dimensions of DOM elements

To get the total size of an element, including the width and height of its content, padding, and borders, and including the space occupied by scrollbars, fetch the element's offsetWidth and offsetHeight properties. These are read-only properties.

[EXAMPLE]

The getBoundingClientRect() method also returns the size of an element, as well as its position relative to the viewport. It was first introduced by Microsoft way back in Internet Explorer 5, but it has since been standardised and works in all of our target browsers.

The method returns a frozen object with six properties: left, top, right, bottom, width, and height. The left, top, right and bottom values are pixel coordinates from the top-left corner of the viewport. The width and height values are the element's dimensions in pixels. Subtracting the left value from the right value would give you the same value as the width property, and subtracting the top from the bottom will give you the same height. All six properties are reported consistently in all modern browsers, but the width and height properties are not reported in legacy editions of Internet Explorer (<9)

var rect = document.getElementById( 'id' ).getBoundingClientRect();

What makes this method so especially handy is that the alternative means of computing an element’s offset position is so horribly slow and buggy that it borders on the comical. The general purpose of this method, or of similar implementations, is to allow the user to get the visual top/left position of an element (such that if they absolutely position an element from the document it would be on top of the desired element – making things like tooltips possible, to name a single application). This means that you have to take into account a huge number of quirks. In no particular order here are some choice bugs which make this very difficult:

Handling table border offsets.

Fixed positioned elements.

Scroll offsets within another element.

Borders of overflowed parent elements.

Miscalculation of absolutely positioned elements.

The getBoundingClientRect() method differs from offsetWidth and offsetHeight in that it takes into account CSS transforms when computing the space occupied by an element. For example an element that is scaled down by 50% – transform: scale(0.5) – will have a rendered size half its layout size. Thus getBoundingClientRect() will not always return values that are comparable to offsetWidth and offsetHeight. The two are not interchangeable.

(jsPerf.com tests reveal that getBoundingClientRect() is consistently faster than getClientRects(). And offsetWidth and offsetHeight are faster still.)

	If you need to know how much space the actual displayed content takes up, including padding but not including the border, margins, or scrollbars, you want to use the clientWidth and clientHeight properties.

	If you need to know the actual size of the content, regardless of how much of it is currently visible, you need to use the scrollWidth and scrollHeight properties. These return the width and height of the entire content of an element, even if only part of it is presently visible due to the use of scroll bars.

	For example, if a 600x400 pixel element is being displayed inside a 300x300 pixel scrollbox, scrollWidth will return 600 while scrollHeight will return 400.

x

http://en.wikipedia.org/wiki/DOM_events

http://www.quirksmode.org/dom/events/

http://www.quirksmode.org/js/events_events.html

https://developer.mozilla.org/en/docs/Web/API/Event

xxx

Mouse events

mousemove

mouse cursor...

The mousemove event works fine, but you should be aware that it may take quite some system time to process all mousemove events. If the user moves the mouse one pixel, the mousemove event fires. Even when nothing actually happens, long and complicated functions take time and this may affect the usability of the site: everything goes very slowly, especially on old computers.

Therefore it’s best to register an onmousemove event handler only when you need it and to remove it as soon as it’s not needed any more:

element.onmousemove = doSomething;

// later

element.onmousemove = null;

mouseover and mouseout

xxx

<ul id="test">

<li>Item 1</li>

<li>Item 2</li>

<li>Item 3</li>

</ul>

xxx

<script>

var test = document.getElementById("test");

// this handler will be executed only once when the cursor moves off 

// the unordered list

test.addEventListener("mouseleave", function( event ) { 

// highlight the mouseleave target

event.target.style.color = "purple";

// reset the color after a short delay

setTimeout(function() {

event.target.style.color = "";

}, 500);

}, false);

// this handler will be executed every time the cursor is moved off a

// different list-item

test.addEventListener("mouseout", function( event ) { 

// highlight the mouseout target

event.target.style.color = "orange";

// reset the color after a short delay

setTimeout(function() {

event.target.style.color = "";

}, 500);

}, false);

</script>

The mouseenter event differs from mouseover in the way it handles event bubbling. If mouseover were used in this example, then when the mouse pointer moved over the Inner element, the handler would be triggered. This is usually undesirable behavior. The mouseenter event, on the other hand, only triggers its handler when the mouse enters the element it is bound to, not a descendant. So in this example, the handler is triggered when the mouse enters the Outer element, but not the Inner element.

mouseover and mouseout are widely considered to be broken because they do bubble and generate many meaningless events for every time the mouse enters a different element within the target element

mouseenter and mouseleave

Microsoft created the proprietary mouseenter and mouseleave events for Internet Explorer 5.5. They resolved the problems with mouseover and mouseout. Because mouseenter and mouseleave do not bubble, they see the entire HTML element to which they're registered as one solid block and do not react the mouseovers and mouseouts taking place on elements nested inside that block. For most usage cases, this behaviour makes mouseenter and mouseleave more convenient than mouseover and mouseout.

	Combined with the behavior of its symmetrical event, mouseenter, the mouseleave DOM Event acts in a very similar way to the CSS :hover pseudo-class.

Although mouseenter and mouseleave were later standardised, it took a long time for other browser vendors to implement them. As of early 2015 they are still missing from Safari and some mobile platforms.

It is possible to mimic the behaviour of mouseenter and mouseleave where they remain unimplemented. But until support is universal, it is better to apply mouseover and mouseout events to elements that have no children. Or just avoid using JavaScript to apply hover effects altogether.

mousedown, mouseup, click

If the user clicks on an element no less than three mouse events fire, in this order:

mousedown, user depresses the mouse button on this element

mouseup, user releases the mouse button on this element

click, one mousedown and one mouseup detected on this element

In general mousedown and mouseup are more useful than click. Some browsers don’t allow you to read out mouse button information onclick. Furthermore, sometimes the user does something with his mouse but no click event follows.

Suppose the user depresses the mouse button on a link, then moves his mouse off the link and then releases the mouse button. Now the link only registers a mousedown event. Similarly, if the user depresses the mouse button, then moves the mouse over a link and then releases the mouse button, the link only registers a mouseup. In neither case does a click event fire.

Whether or not this is a problem depends on the user interaction you want. But you should generally register your script onmousedown/up, unless you’re completely sure you want the click event and nothing else.

If you use alerts in handling these events, the browser may lose track of which event took place on which element and how many times it took place, messing up your scripts. So it’s best not to use alerts.

Dblclick

The dblclick event is rarely used. Even when you use it, you should be sure never to register both an onclick and an ondblclick event handler on the same HTML element. Finding out what the user has actually done is nearly impossible if you register both.

After all, when the user double–clicks on an element one click event takes place before the dblclick. Besides, in Netscape the second click event is also separately handled before the dblclick. Finally, alerts are dangerous here, too.

So keep your clicks and dblclicks well separated to avoid complications.

target and relatedTarget

IE's srcElement is deprecated

W3C added the relatedTarget property to mouseover and mouseout events. This contains the element the mouse came from in case of mouseover, or the element it goes to in case of mouseout.

Microsoft created two properties to contain this information:

fromElement refers to the element the mouse comes from. This is interesting to know in case of mouseover.

toElement refers to the element the mouse goes to. This is interesting to know in case of mouseout.

Keyboard events

In all browsers, key strokes can be captured and handled. But handling them consistently is not at all easy. The problem is that there was no attempt to try to standardise these events until DOM 3. As a result, there is still a lot of inconsistent behaviour from one browser to another. 

Below I explain how keyboard events should be handled to ensure consistent reporting across all target browsers.

http://www.quirksmode.org/dom/events/keys.html

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which

https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/charCode#Notes

http://api.jquery.com/event.which/

http://stackoverflow.com/questions/4471582/javascript-keycode-vs-which

http://stackoverflow.com/questions/3050984/javascript-event-e-which

Keyboard event targets

Keyboard events can be attached to the whole document, or to any element within it. In the case of individual elements, keyboard events will be fired only when the element also has focus. Keyboard events attached to the document will be fired when the document's window is the active window.

Attaching keyboard events to window works the same as attaching them to window.document in all modern browsers, but this does not work in legacy editions of Internet Explorer (<9).

Keyboard event types

There are three main keyboard events: keyup, keypress, and keydown.
•keydown fires when the user presses a key.
•keypress fires just before a key performs its default action, e.g. inserting a character.
•keyup fires when the user releases a key.

The keypress event continues to fire repeatedly while a key is held down. That makes sense. Holding down an alphanumeric key when a text box is in focus will usually result in that character being typed over and over again. 

Auto-repeat behaviour is inconsistent for keydown and keyup events: keydown events are supposed to auto-repeat, but this is not the case for every key in every browser; keyup events are not supposed to auto-repeat, yet this behaviour has been reported in Firefox on some Linux distros. We need to consider this inconsistent behaviour when authoring callback functions for keydown and keyup events.

Identifying keys and characters

Working out which key is pressed is tricky, but not impossible. The first thing to be clear about is whether you want to identify what physical key was pressed, or what character (if any) was typed as the result of pressing the key. These are different things and must be handled in different ways. 

If you want to know what key was pressed, listen for keydown or keyup events. Using keydown will give you the opportunity to cancel the key's default action, but keyup will not. 

If you want to know what character was produced, listen for keypress events. You can also cancel a key's default action from inside a keypress callback.

Never use keypress to try to identify physical keys pressed. You will not succeed. There are just too many keyboard formats to contend with. For example, pressing the "4" key while depressing Shift typically produces a dollar sign character when using English-language keyboards. But the same key combination may produce other characters on other keyboards.

Never use keypress events to listen for control keys such as Escape and ArrowLeft. (Control keys are any key that does not produce a physical character.) Firefox and Opera are the only two browsers that fire keypress events in this scenario, and they are not supposed to. This behaviour is non-standard and should be ignored by authors.

A small number of control keys produce printable output: Tab, Enter, and Spacebar. All modern browser do in fact fire keypress events when Enter and Spacebar are pressed, but only Gecko browsers do so when the Tab key is pressed. Due to this inconsistency, it is recommended that keydown or keyup be used to detect every control key, even the ones that produce printable characters. Use keypress for everything else.

To find out what key was pressed, or what character resulted, we check properties that are attached to the KeyBoardEvent object, which is provided to our callback functions. There are several properties that each do slightly different things: keyCode, charCode, which, and key.

The KeyboardEvent.keyCode property returns a numeric code that represents the physical key that was pressed. This is the property that you want to query from inside keydown and keyup callbacks. Although we can use KeyboardEvent.keyCode to check if an alphanumeric key – "5" or "E", for example – was pressed, better to use keydown/keyup events in conjunction with the keyCode property to listen exclusively for control keys – anything that is non-printable and available on most localised keyboards. I recommend that you limit yourself to listening out only for the control keys listed below (their key codes are in brackets). Do not listen out for the function keys (F1-F25), Caps Lock or Num Lock, or for OS-specific keys such as the Windows and Apple keys. These keys do not exist on every keyboard and often their default behaviour cannot be disabled.
•Backspace (8)
•Tab (9)
•Enter (13)
•Escape (27)
•Spacebar (32)
•PageUp (33)
•PageDown (34)
•End (35)
•Home (36)
•ArrowLeft (37)
•ArrowUp (38)
•ArrowRight (39)
•ArrowDown (40)
•Delete (46)

The KeyboardEvent.charCode property returns the Unicode value of a character key following a keypress event. It takes into account whether the Shift key was depressed or if Caps Lock was on when the key was pressed. You can easily convert the Unicode value to an actual character with String.fromCharCode().

String.fromCharCode( e.charCode )

In some browsers the KeyboardEvent.keyCode property is sometimes made available to keypress callback functions. And in some browsers the KeyBoardEvent.charCode property is sometimes made available to keydown and keyup callbacks. This is non-standard behaviour, and anyway it is inplemented inconsistently between browsers. So, restrict yourself to checking only the keyCode property for keydown and keyup events, and checking only the charCode property for keypress events. 

When implementing shortcut key handlers (e.g. capital "S" to save), listen for keypress events and check the KeyBoardEvent.charCode property. When implementing keyboard controls (e.g. ArrowLeft and ArrowRight to toggle the visiblility of panels in a tabbed interface), listen for keydown events and check the KeyboardEvent.keyCode property.

The KeyboardEvent.which property attempts to normalize KeyboardEvent.keyCode and KeyboardEvent.charCode. If the key produces an alphanumeric character, KeyboardEvent.which will return the character code, otherwise it will return the key code. The KeyboardEvent.which property seems to work consistently in every browser. Use it, but fallback to KeyboardEvent.keyCode and KeyboardEvent.charCode just in case.

var char_code = event.charCode || event.charCode;

All of these properties – KeyboardEvent.keyCode, KeyboardEvent.charCode, and KeyboardEvent.which – have been dropped from current web standards, and support may well be removed from browsers entirely in the future. The plan is to consolidate their behaviour around a single new property, KeyboardEvent.key. It is meant to be available for every type of keyboard event – keydown, keypress, and keyup – and should report consistent values in every case. If the key strokes result in printed output, KeyboardEvent.key should report that output verbatim. It will thus distinguish case, "a" and "A". For control keys, KeyboardEvent.key should return a consistent name for that key, such as "Tab" and "Enter". If the key cannot be identified, "Unidentified" will be returned.

Unfortunatately, (partial) support for event.key is currently limited to Internet Explorer (9+) and Firefox (23+). The implementation does not quite match the current specification, because some of the keys have been subsequently renamed. For example, Firefox reports spaces as "Spacebar" when the spec says it should be " ". And "Esc" has been renamed "Escape", and "Del" is meant to be "Delete". These problems can be easily accommodation, though.

It will be sometime before the excellent KeyboardEvent.key property works consistently universally. But it will be even longer before keyCode, charCode and which disappear entirely. They might not be standards, but they're all we've got for now.

Andy Earnshaw has written a polyfill for KeyboardEvent.key, though it needs some updating to adhere to the current specification:
https://gist.github.com/andyearnshaw/11333894#file-key-polyfill-js

Incidentally, although Google Chrome and Safari do not yet support KeyboardEvent.key, they do support KeyboardEvent.keyIdentifier. This property was defined in an old draft of DOM Level 3 Events, but ultimately dropped in favour of KeyboardEvent.key. The KeyboardEvent.keyIdentifier property is a little idiosyncratic. For most alphanumeric characters it seems to return a Unicode value, such as "U+0041" for "a" and "A" (it does not differentiate case). But sometimes it will return odd named keys, such as "Help" when "/" is typed on English language keyboards. Best to leave KeyboardEvent.keyIdentifier alone. It will probably get quietly dropped from WebKit browsers soon, anyway.

Input events

The input event is supported by all browsers. It is fired immediately after a change is made to a <textarea> or freeform text <input> field. Typically, this is immediately after a keypress event. 

Unlike the keypress event, the input event does not give any information about what key was pressed or what character was typed. All you can retrieve is the new, full value for the text box. And, unlike the keypress event, the input event works only for text input controls.

A cancelled keypress event may prevent a subsequent input event from firing.

Some browsers previously supported an event called textInput ("textinput" in IE). Originally it had been intended to replace the keypress event, but this idea was abandoned. Do not use it. It is non-standard, it doesn't work in Firefox or Opera, and support may be dropped from other browsers.

xxx

onblur vs onchange - they are not always fired in the same order:

In Firefox, tabbing out of a changed field will fire onchange then onblur, and it will normally do the same in IE. However, if you press the enter key instead of tab, in Firefox it will fire onblur then onchange, while IE will usually fire in the original order. However, I've seen cases where IE will also fire blur first, so be careful. You can't assume that either the onblur or the onchange will happen before the other one.

Also, the change event is not always fired consistently from one browser to another: http://www.quirksmode.org/dom/events/change.html. For radios and checkboxes, you might prefer to check for 'click' events, for better backwards compatibility with IE8-.

Form onsubmit

Programmatically submitting a form – form.submit() – will bypass any onsubmit events established for the form: http://stackoverflow.com/questions/18774821/

HTML event attributes

Do not use HTML event attributes, onclick, oninput, onmouseout, etc. Prefer to unobtrusively enhance web interfaces with the EventTarget.addEventListener API.

But if you must use HTML event attributes, be sure fully qualify global variables and functions against the window object. So instead of this:

<div onclick="myCallback()">Click here</div>

Write this:

<div onclick="window.myCallback()">Click here</div>

The purpose is to reduce name collisions. When functions and variables are written inside event attributes like this, JavaScript first looks for them on the element itself, then on the element's parent form if it has one, then on the document, and then finally on the global window object. That provides plenty of opportunity for the wrong thing to be referenced. This may happen unexpectedly at any time in the future when native APIs are extended with new functionality.

Triggering events dynamically

When you change the input value of textbox dynamically, for example, no onchange events will be fired as though the user changed the value.

input.value = 'New value';

To trigger any onchange events that may be assigned to the element, use to UIEvent API.

var event = new UIEvent("change", {

"view": window,

"bubbles": true,

"cancelable": true

});

input.dispatchEvent(event);

Use the MouseEvent interface to simulate clicks:

var event = new MouseEvent('click', {

'view': window,

'bubbles': true,

'cancelable': true

});

var cb = document.getElementById('checkbox'); 

var canceled = !cb.dispatchEvent(event);

if (canceled) {

// A handler called preventDefault.

alert("canceled");

} else {

// None of the handlers called preventDefault.

alert("not canceled");

}

More: http://www.2ality.com/2013/06/triggering-events.html

Creating custom events

Use new Event to create custom events... sadly this does not work in Internet Explorer, but there is an older API that you can use to polyfill IE. See:

https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events

https://developer.mozilla.org/en-US/docs/Web/API/Event/Event

xxx

Element.scrollLeft and Element.scrollTop

These were proprietary properties developed by Microsoft for its Internet Explorer, but they were subsequently adopted by all other browsers and became a de facto standard. They are now on a standardisation track.

These properties return the number of pixels that an element is scrolled from the left and from the top. If the element cannot be scrolled – if it does not generate a scrollbar – the return values will always be 0. Both properties enjoy a long history and they work reliably in every target browser. But be careful; these properties must not be used to get the scroll distance of the window, the document, or even for document.body. See the notes below on window.scrollX and window.scrollY.

window.scrollX and window.scrollY

The window.scrollX and window.scrollY properties should return the number of pixels that the document has been scrolled horizontally and vertically, respectively. These are new properties, being defined in the draft "CSS Object Model View Module" specification (http://dev.w3.org/csswg/cssom-view/). They are not yet supported by all target browsers, so fallback to the older window.pageXOffset and window.pageYOffset properties, which are universally supported.

var scrollY = window.scrollY || window.pageYOffset;

window.scrollMaxX and window.scrollMaxY

The window.scrollMaxX and window.scrollMaxY properties should return the maximum number of pixels that the document can be scrolled horizontally and vertically. These are new also, and still return undefined in several browsers. You can fallback to document.body.scrollWidth and document.body.scrollHeight.

var max_scroll_x = window.scrollMaxX || document.body.scrollWidth;

More: http://www.quirksmode.org/dom/w3c_cssom.html

window.scrollBy

Here's a DOM method that's really simple, is supported in all browsers including oldIE, but doesn't seem to have been added to the spec until recently: window.scrollBy().

Let's look at some example code. We'll assume we have a really long page of content with two button elements:

var btn = document.querySelector('button');

 btn2 = document.querySelectorAll('button')[1];

btn.addEventListener('click', function () {

 window.scrollBy(0, window.innerHeight);

}, false);

btn2.addEventListener('click', function () {

 window.scrollBy(0, -(window.innerHeight));

}, false);

If you click the first button, the page scrolls down. If you click the second button, it scrolls up. If there's no room to scroll, it will do nothing. The scrollBy() method takes two arguments, an number that's a pixel value to scroll horizontally (x-axis) and one to scroll vertically (y-axis, which is what you'll normally use).

******

<canvas>

The <canvas> element is supported by all modern desktop and mobile platforms. Internet Explorer was the last to implement it, with the release of version 9 in 2011.

Every <canvas> element must have id, width, and height attributes. Identifiers – each of which must be unique in the document – should be prefixed "canvas-" to reduce the chances of collisions with other things in the document.

<canvas id="canvas-example" width="200" height="200"></canvas>

Strictly speaking, <canvas> elements are not part of the Document Object Model. And therefore they are not interpreted as being part of the document's content. Screen readers and other assistive technologies will generally ignore them. The HTML Accessibility Task Force (http://www.w3.org/WAI/PF/html-task-force) at the W3C is brainstorming how to make canvas content more accessible. In the meantime we should provide alternative material to users who cannot access canvas content. For example if a <canvas> element is used to generate a graph, consider displaying a basic table of the raw data. Just include it inside the <canvas> element itself. (Conveniently, JavaScript could extract the original data from the table to generate the graph.)

<canvas id="canvas-exam-results" width="250" height="500">

<table>...</table>

</canvas>

WebGL

WebGL extends the <canvas> element to allow for the dynamic rendering of 3D graphics in the browser. It has been used in the MapsGL version of Google Maps since 2011, when it was also introduced to the Google Chrome browser. It has also been supported by Mozilla Firefox since version 4.0, and Apple Safari since version 6.0. However WebGL remains disabled by default in Opera 11/12, is only partically implemented in Internet Explorer 11, and support is patchy in the native Android browser and iOS Safari. So, like Google Maps, WebGL may be rolled out only as a progressive enhancement for browsers that support it. Provide plain old 2D graphics by default.

<svg>

Embedded scalable vector graphics are another standard – which predates <canvas> – for drawing shaped in browsers. But while canvas is raster-based, SVGs are vectors. This makes it a bit easier for browsers to re-render SVG objects when their attributes change. For this reason, prefer to use <svg> for things like data visualisation.

The XMLHttpRequest interface allows JavaScript to make HTTP requests. It means we can get our web applications to post data to, and fetch data from, other resources without needing to reload the entire web page. This process is commonly known as Ajax, which is short for "asynchronous JavaScript and XML". Despite the reference to XML in the name, we can actually use XMLHttpRequest to retrieve data in any format. In fact, it is more common to use JSON as a data interchange format, as it is easier for JavaScript to parse.

The XMLHttpRequest property is available on the global window object.

var xhr = new XMLHttpRequest();

In general: use Ajax only for data. Not for returned new views, and partials. All of the markup you should ever need should be embedded in the main document when it is first loaded.

Simple example

In the following example, an asynchronous request is made and an onreadystatechange event handler processes the data returned.

var xhr = new XMLHttpRequest();

xhr.open( 'GET', "http://localhost/test.json", true );

xhr.onreadystatechange = function () {

if ( xhr.readyState === 4 /* complete */ ) {

if ( xhr.status === 200 || xhr.status === 304 ) {

console.log( xhr.responseText );

}

}

};

xhr.send();

http://www.sitepoint.com/forums/showthread.php?570588-for-AJAX-is-the-status-200-and-304-both-needed-to-be-considered-for-a-success

Properties

The following XMLHttpRequest properties are supported by all target browsers.

readyState

Returns the current state of the XMLHttpRequest.

response

Returns the response entity body, which is the fragment of the entity body of the response received so far (if LOADING) or the complete entity body of the response (if DONE).

responseText

Returns the text response entity body, a string representing the response entity body, which is the fragment of the entity body of the response received so far (if LOADING) or the complete entity body of the response (if DONE).

responseType

Returns or sets the format the response will be returned in.

responseXML

Returns the response to the request as a DOM Document object, or null if the request was unsuccessful, has not yet been sent, or cannot be parsed as XML or HTML.

status

Returns the HTTP result code (status) of the response to the request.

statusText

Returns the HTTP status text.

timeout

Returns or sets the number of milliseconds a request can take before automatically being terminated.

upload

Returns the associated XMLHttpRequestUpload object. It can be used to gather transmission information when data is transferred to a server.

withCredentials

Returns or sets whether cross-site Access-Control requests should be made using credentials such as cookies or authorization headers.

Methods

The following XMLHttpRequest methods are supported by all target browsers:

abort

Stops an asynchronous XMLHttpRequest in progress.

getAllResponseHeaders

Returns all the response headers as a string, or null if no response has been received.

getResponseHeader

Returns the string containing the text of the specified header, or null if the response has not been received or the header does not exist.

open

Initializes an XMLHttpRequest.

overrideMimeType

Overrides the MIME type returned by the server.

send

Initiates the request defined by the XMLHttpRequest.

setRequestHeader

Sets the value of an XMLHttpRequest header.

Examples:

xhr.setRequestHeader('X-Requested-With', 'XMLHttpRequest');

xhr.setRequestHeader('Content-Type', content_type);

// Chrome/Safari throw errors if you try to manually set these headers,

// because the browser always uses its own values. Messing around with

// these could expose various "request smuggling" attacks:

// Ref: http://stackoverflow.com/questions/7210507/

//xhr.setRequestHeader('Content-length', data.length);

//xhr.setRequestHeader('Connection', 'close');

Events

The following XMLHttpRequest events are supported by all target browsers.

abort

When the request has been aborted. For instance, by invoking the abort() method.

error

When the request has failed.

load

When the request has successfully completed.

loadend

When the request has completed (either in success or failure).

loadstart

When the request starts.

progress

While sending and loading data.

readystatechange

Fires whenever the readyState of the request changes. Mostly used to determine whether the body of the response is available for handling.

http://stackoverflow.com/questions/16848955/

timeout

When the author specified timeout has passed before the request could complete.

XMLHttpRequest Level 2

XMLHttpRequest Level 2 is a major upgrade to the original XMLHttpRequest standard. It adds even more properties, methods and events to the XMLHttpRequest interface. New tricks include the responseType property, for declaring the expected file format of the response body; Cross Origin Resource Sharing (CORS) for cross-domain Ajax requests; and the onload function, a sort of shortcut for onreadystatechange/readyState === 4:

/* XMLHttpRequest: */

xhr.onreadystatechange = function () {

if ( xhr.readyState === 4 /* complete */ ) {

if ( xhr.status === 200 ) {

console.log( xhr.responseText );

}

}

};

/* XMLHttpRequest Level 2: */

xhr.onload = function() { 

if ( xhr.status === 200 ) {

console.log( this.responseText );

}

}

XMLHttpRequest Level 2 is fully implemented in all target browsers and many legacy browsers too, with one notable exception... Internet Explorer 9. Until we drop support for this browser, none of the XMLHttpRequest Level 2 APIs can be used. For now, restrict yourself to the properties, methods and events listed above.

Reference: http://www.w3.org/TR/XMLHttpRequest2/

Cross-domain requests

Almost every browser on all desktop and mobile platforms supports cross-origin requests via XMLHttpRequest. This feature is one of a bundle of new goodies to be mixed into XMLHttpRequest Level 2. Internet Explorer came late the party, though, adding support for XMLHttpRequest Level 2 only in version 10. But IE 8/9 support a proprietary interface for cross-domain requests which we can use as a fallback for those browsers.

You can't request stuff from any old domain, though. The destination server must be configured to support Cross-Origin Resource Sharing (CORS). A header in the response, called Access-Control-Allow-Origin, determines which domains are allowed to make requests. 

For example, imagine you have an application that lives on www.example.com and you need to pull data from api.example.com. Normally if you make an Ajax call from the first to the second domain, the request would fail and the browser would throw an origin mismatch error. To allow requests from www.example.com, the server for api.example.com would need to return the following header:

Access-Control-Allow-Origin: http://www.example.com

If your data is public, you will need to allow any domain to make a request:

Access-Control-Allow-Origin: *

As long as the remote server has enabled CORS, how you make a remote Ajax request is basically no different to a local one. But you do need to account for the remote server possibly not supporting CORS (or removing support in the future without notice). And we also need a workaround for Internet Explorer 9, which is not capable of making cross-origin Ajax calls. Here's how we do that:

function remoteAjax( method, url ) {

var xhr = new XMLHttpRequest();

if ( 'withCredentials' in xhr ) {

xhr.open( method, url, true ); 

} 

else if ( typeof XDomainRequest !== 'undefined' ) {

/* IE8/9: */

xhr = new XDomainRequest();

xhr.open( method, url );

} 

else {

/* CORS not supported: */

xhr = null;

}

return xhr;

}

Usage example:

var url = 'api.example.com';

var xhr = remoteAjax( 'GET', url );

if ( xhr ) {

xhr.onload = function() {

var text = xhr.responseText;

//...

};

xhr.onerror = function() {

//...

};

xhr.send();

}

XDomainRequest is an implementation of HTTP access control (CORS) that worked in Internet Explorer 8 and 9. It was removed in Internet Explorer 10 in favor of using XMLHttpRequest with proper CORS; if you are targeting Internet Explorer 10 or later, or wish to support any other browser, you need to use standard HTTP access control. ... The XDomainRequest interface can send both GET and POST requests, but not other types of request.

if(window.XDomainRequest){

 var xdr = new XDomainRequest();

 xdr.open("get", "http://example.com/api/method");

 xdr.onprogress = function () {

 //Progress

 };

 xdr.ontimeout = function () { 

 //Timeout

 };

 xdr.onerror = function () { 

 //Error Occured

 };

 xdr.onload = function() {

 //success(xdr.responseText);

 }

 setTimeout(function () {

 xdr.send();

 }, 0);

}

Note: The xdr.send() call is wrapped in a timeout (see window.setTimeout() to prevent an issue with the interface where some requests are lost if multiple XDomainRequests are being sent at the same time.

The XDomainRequest is built to be secure in multiple ways.

The origin's security protocol must match that of the requested URL. (http to http, https to https). If these do not match, the request will error "Access is Denied".

The requested URL's server must have the Access-Control-Allow-Origin header set to either all ("*") or to include the origin of the request.

This interface and its methods are non-standard.

***

Dynamic file uploads

The introduction of XMLHttpRequest was a huge leap forward in the development of the web platform. The ability to communicate with a web server without reloading the page has revolutionised how web applications are built.

	XMLHttpRequest Level 2 adds support for the new FormData interface. FormData objects provide a way to easily construct a set of key/value pairs representing form fields and their values, which can then be easily sent using the XMLHttpRequest send() method.

	A nice feature added to XMLHttpRequests in recent years is the ability to handle file uploads. Traditionally many developers have resorted to using technologies like Flash to upload files to a server. The problem with this approach is that the user needs to have a third-party browser plugin installed.

***

... Instead, we can use FormData, another new data type conceived for XHR2. FormData is convenient for creating an HTML <form> on-the-fly, in JavaScript. That form can then be submitted using AJAX:

function sendForm() {

 var formData = new FormData();

 formData.append('username', 'johndoe');

 formData.append('id', 123456);

 var xhr = new XMLHttpRequest();

 xhr.open('POST', '/server', true);

 xhr.onload = function(e) { ... };

 xhr.send(formData);

}

Essentially, we're just dynamically creating a <form> and tacking on <input> values to it by calling the append method.

Of course, you don't need to create a <form> from scratch. FormData objects can be initialized from and existing HTMLFormElement on the page. For example:

<form id="myform" name="myform" action="/server">

 <input type="text" name="username" value="johndoe">

 <input type="number" name="id" value="123456">

 <input type="submit" onclick="return sendForm(this.form);">

</form>

function sendForm(form) {

 var formData = new FormData(form);

 formData.append('secret_token', '1234567890'); // Append extra data before send.

 var xhr = new XMLHttpRequest();

 xhr.open('POST', form.action, true);

 xhr.onload = function(e) { ... };

 xhr.send(formData);

 return false; // Prevent page from submitting.

}

An HTML form can include file uploads (e.g. <input type="file">) and FormData can handle that too. Simply append the file(s) and the browser will construct a multipart/form-data request when send() is called:

function uploadFiles(url, files) {

 var formData = new FormData();

 for (var i = 0, file; file = files[i]; ++i) {

 formData.append(file.name, file);

 }

 var xhr = new XMLHttpRequest();

 xhr.open('POST', url, true);

 xhr.onload = function(e) { ... };

 xhr.send(formData); // multipart/form-data

}

document.querySelector('input[type="file"]').addEventListener('change', function(e) {

 uploadFiles('/server', this.files);

}, false);

Response types

Fetching a file as a binary blob has been painful with XHR. Technically, it wasn't even possible. One trick that has been well documented involves overriding the mime type with a user-defined charset as seen below.

	The old way to fetch an image:

var xhr = new XMLHttpRequest();

xhr.open('GET', '/path/to/image.png', true);

// Hack to pass bytes through unprocessed.

xhr.overrideMimeType('text/plain; charset=x-user-defined');

xhr.onreadystatechange = function(e) {

 if (this.readyState == 4 && this.status == 200) {

 var binStr = this.responseText;

 for (var i = 0, len = binStr.length; i < len; ++i) {

 var c = binStr.charCodeAt(i);

 //String.fromCharCode(c & 0xff);

 var byte = c & 0xff; // byte at offset i

 }

 }

};

xhr.send();

While this works, what you actually get back in the responseText is not a binary blob. It is a binary string representing the image file. We're tricking the server into passing the data back, unprocessed. Even though this little gem works, I'm going to call it black magic and advise against it. Anytime you resort to character code hacks and string manipulation for coercing data into a desirable format, that's a problem.

	In the previous example, we downloaded the image as a binary "file" by overriding the server's mime type and processing the response text as a binary string. Instead, let's leverage XMLHttpRequest's new responseType and response properties to inform the browser what format we want the data returned as.

xhr.responseType

Before sending a request, set the xhr.responseType to "text", "arraybuffer", "blob", or "document", depending on your data needs. Note, setting xhr.responseType = '' (or omitting) will default the response to "text".

xhr.response

After a successful request, the xhr's response property will contain the requested data as a DOMString, ArrayBuffer, Blob, or Document (depending on what was set for responseType.)

With this new awesomeness, we can rework the previous example, but this time, fetch the image as an Blob instead of a string:

var xhr = new XMLHttpRequest();

xhr.open('GET', '/path/to/image.png', true);

xhr.responseType = 'blob';

xhr.onload = function(e) {

 if (this.status == 200) {

 // Note: .response instead of .responseText

 var blob = new Blob([this.response], {type: 'image/png'});

 ...

 }

};

xhr.send();

xxx

To review:

http://www.javascripture.com/FormData

https://developer.mozilla.org/en-US/docs/Web/Guide/Using_FormData_Objects

http://blog.teamtreehouse.com/uploading-files-ajax

https://developer.mozilla.org/en/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest

https://developer.mozilla.org/en/docs/Web/API/FileList

This is the old way of mimicing ajaxy file uploads - submit a form to a hidden iframe. This works, but it's difficult to get any kind of meaningful feedback. Better solution for file uploads is progressive enhancement -use a standard POST for IE9, but enhance with XMLHttpRequest 2 for other browsers.

http://www.openjs.com/articles/ajax/ajax_file_upload/
